File: java/AlgorithmAndDataStructure/src/test/java/algorithms/BinarySearchTest.java
package algorithms;

import junit.framework.TestCase;

import java.util.ArrayList;
import java.util.Arrays;

public class BinarySearchTest extends TestCase {
	private ArrayList<Integer> a;

	public void setUp() throws Exception {
		super.setUp();
		a = new ArrayList<Integer>(Arrays.asList(1, 2, 3, 4));
	}

	public void testLowerBound() throws Exception {
		assertEquals(2, BinarySearch.lowerBound(a, 3));
		assertEquals(0, BinarySearch.lowerBound(a, 1));
		assertEquals(4, BinarySearch.lowerBound(a, 5));
	}

	public void testUpperBound() throws Exception {
		assertEquals(3, BinarySearch.upperBound(a, 3));
		assertEquals(4, BinarySearch.upperBound(a, 4));
	}

	public void testBinarySearch() throws Exception {
		assert BinarySearch.binarySearch(a, 3);
		assert !BinarySearch.binarySearch(a, 0);
		assert !BinarySearch.binarySearch(a, 5);
	}

	public void testEqualRange() throws Exception {
		Pair equalThree = BinarySearch.equalRange(a, 3);
		assertEquals(2, equalThree.getFirst());
		assertEquals(3, equalThree.getSecond());
	}
}

File: php/BinarySearchTest.php
<?php
namespace Algorithm;

class BinarySearchTest extends \PHPUnit_Framework_TestCase
{
    private $ary;

    public function setUp() {
        // Required a sorted array
        $this->ary = [2, 3, 5, 8, 13];
    }

    public function testLowerBound()
    {
        // Return Null if the upper bound is out of range
        $this->assertNull(lowerBound($this->ary, 14));
        $this->assertEquals(4, lowerBound($this->ary, 13));
        $this->assertEquals(3, lowerBound($this->ary, 7));
        $this->assertEquals(1, lowerBound($this->ary, 3));
        $this->assertEquals(0, lowerBound($this->ary, 2));
        $this->assertEquals(0, lowerBound($this->ary, 1));
    }

    public function testUpperBound()
    {
        $this->assertNull(upperBound($this->ary, 13));
        $this->assertEquals(2, upperBound($this->ary, 3));
        $this->assertEquals(1, upperBound($this->ary, 2.7));
        $this->assertEquals(0, upperBound($this->ary, 1));
    }

    public function testBinarySearch()
    {
        $this->assertFalse(binarySearch($this->ary, 16));
        $this->assertTrue(binarySearch($this->ary, 13));
        $this->assertTrue(binarySearch($this->ary, 8));
        $this->assertFalse(binarySearch($this->ary, 6));
        $this->assertFalse(binarySearch($this->ary, 1));
    }

    public function testEqualRange()
    {
        $this->assertEquals(equalRange($this->ary, 2), [0, 1]);
        $this->assertEquals(equalRange($this->ary, 4), [2, 2]);
        $this->assertEquals(equalRange($this->ary, 5), [2, 3]);
        $this->assertEquals(equalRange($this->ary, 13), [4, null]);
    }
}
?>
